### PR TITLE
Pvhome #93: Realizar o desenvolvimento do repositório NodeJS JDBC

### PR DESCRIPTION
**Pull request by pvpmartins**

### CODE REVIEW

File: src/callablestatement.ts [hunk 1]
```diff
@@ -23,8 +23,8 @@ export class CallableStatement extends PreparedStatement {
   }
 
   async getBigDecimal(arg1: number | string): Promise<any> {
```
Comment (pontusvision) [inline src/callablestatement.ts]: let's try to type this correctly; avoid Promise<any>

File: src/callablestatement.ts [hunk 2]
```diff
@@ -36,8 +36,8 @@ export class CallableStatement extends PreparedStatement {
   }
 
   async getBlob(arg1: number | string): Promise<any> {
-    if (typeof arg1 === 'number' || typeof arg1 === 'string') {
-      return new Promise((resolve, reject) => {
+    if (typeof arg1 === "number" || typeof arg1 === "string") {
+      return new Promise<any>((resolve, reject) => {
```
Comment (pontusvision) [inline src/callablestatement.ts]: let's try to type this correctly; avoid Promise<any>

File: src/databasemetadata.ts [hunk 3]
```diff
@@ -1,95 +1,3593 @@
...
+        (err: Error, result: any) => {
+          if (err) {
+            return reject(err);
+          }
+          resolve(new ResultSet(result));
+        }
+      );
+    });
+  }
+  async allProceduresAreCallable(): Promise<any> {
```
Comment (pontusvision) [inline src/databasemetadata.ts]: let's try to type this correctly; avoid Promise<any>

File: src/pool.ts [hunk 4]
```diff
@@ -1,180 +1,252 @@
...
 
   async status(): Promise<any> {
-    const status = {
-      available: this._pool.length,
-      reserved: this._reserved.length,
-      pool: this.connStatus([], this._pool),
-      rpool: this.connStatus([], this._reserved)
-    };
-
+    const status: any = {};
```
Comment (pontusvision) [inline src/pool.ts]: let's try to type this correctly; avoid Promise<any>

File: src/pool.ts [hunk 5]
```diff
@@ -1,180 +1,252 @@
...
+    status.available = this._pool.length;
+    status.reserved = this._reserved.length;
+    status.pool = this.connStatus([], this._pool);
+    status.rpool = this.connStatus([], this._reserved);
     return status;
   }
 
+  private connStatus(acc: any[], pool: any[]): any[] {
```
Comment (pontusvision) [inline src/pool.ts]: let's try to type this correctly; avoid any

File: src/pool.ts [hunk 6]
```diff
@@ -1,180 +1,252 @@
...
+      jinst.events.emit("initialized");
+    } catch (err) {
+      winston.error(err);
+      throw err; // Rethrow the error for handling in the calling code
     }
-
-    jinst.events.emit('initialized');
   }
 
   async reserve(): Promise<any> {
```
Comment (pontusvision) [inline src/pool.ts]: let's try to type this correctly; avoid any

File: src/pool.ts [hunk 7]
```diff
@@ -185,48 +257,55 @@ class Pool {
...
     }
 
-    this._closeIdleConnectionsInArray(this._pool, this._maxidle);
-    this._closeIdleConnectionsInArray(this._reserved, this._maxidle);
+    this.closeIdleConnectionsInArray(this._pool, this._maxidle);
+    this.closeIdleConnectionsInArray(this._reserved, this._maxidle);
   }
 
-  private _closeIdleConnectionsInArray(array: any[], maxIdle: number): void {
+  private closeIdleConnectionsInArray(array: any[], maxIdle: number): void {
```
Comment (pontusvision) [inline src/pool.ts]: let's try to type this correctly; avoid any

File: src/pool.ts [hunk 8]
```diff
@@ -185,48 +257,55 @@ class Pool {
...
         if (conn.lastIdle < maxLastIdle) {
-          conn.conn.close(() => { });
+          conn.conn.close((err: Error) => {});
           array.splice(i, 1);
         }
       }
     }
   }
 
   async release(conn: any): Promise<void> {
```
Comment (pontusvision) [inline src/pool.ts]: let's try to type this correctly; avoid any

File: src/resultset.ts [hunk 9]
```diff
@@ -1,5 +1,3 @@
-/* jshint node: true */
-"use strict";
 import _ from 'lodash';
```
Comment (pontusvision) [inline src/resultset.ts]: let's try to get rid of lodash

File: src/statement.ts [hunk 10]
```diff
@@ -1,14 +1,22 @@
 /* jshint node: true */
 "use strict";
 
-import _ from 'lodash';
-import ResultSet from './resultset';
-import jinst from './jinst';
+import _ from "lodash";
```
Comment (pontusvision) [inline src/statement.ts]: Let's get rid of lodash

File: src/callablestatement.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
-import _ from 'lodash';
-import { PreparedStatement } from './preparedstatement';
+import _ from "lodash";
```
Comment (pontusvision) [inline src/callablestatement.ts]: Please remove

File: src/connection.ts [hunk 12]
```diff
@@ -1,111 +1,589 @@
-import _ from 'lodash';
-import { jinst } from './jinst';
-import { CallableStatement } from './callablestatement';
-import { PreparedStatement } from './preparedstatement';
-import { DatabaseMetaData } from './databasemetadata';
-import { Statement } from './statement';
-import { SQLWarning } from './sqlwarning';
+import _ from "lodash";
```
Comment (pontusvision) [inline src/connection.ts]: Please remove lodash

File: src/databasemetadata.ts [hunk 13]
```diff
@@ -1,95 +1,2387 @@
...
-import { jinst } from './jinst';
+import ResultSet from "./resultset.js";
+import Connection from "./connection.js";
+import jinst from "./jinst.js";
 
 const java = jinst.getInstance();
 
-export class DatabaseMetaData {
+// Assuming these types are defined elsewhere in your codebase
+type Callback<T> = (err: Error | null, result: T) => void;
```
Comment (pontusvision) [inline src/databasemetadata.ts]: Remove the Callback<T> type.

File: src/drivermanager.ts [hunk 14]
```diff
@@ -1,93 +1,131 @@
-import _ from 'lodash';
-import { jinst } from "./jinst";
+import _ from "lodash";
```
Comment (pontusvision) [inline src/drivermanager.ts]: Remove Lodash

File: src/drivermanager.ts [hunk 15]
```diff
@@ -1,93 +1,131 @@
...
+import jinst from "./jinst.js";
```
Comment (pontusvision) [inline src/drivermanager.ts]: remove .js?

File: src/jinst.ts [hunk 16]
```diff
@@ -23,19 +23,19 @@ const jinst: Jinst = {
     if (!isJvmCreated() && option) {
       java.options.push(option);
     } else if (isJvmCreated()) {
-      winston.error("You've tried to add an option to an already running JVM!");
-      winston.error("This isn't currently supported. Please add all option entries before calling any java methods");
-      winston.error("You can test for a running JVM with the isJvmCreated function.");
+      console.error("You've tried to add an option to an already running JVM!");
```
Comment (pontusvision) [inline src/jinst.ts:76]: Change the jinst to a class and return it  instead?

File: src/drivermanager.ts [hunk 17]
```diff
@@ -1,93 +1,132 @@
...
-      (_.isObject(args[1]) && !args[2])
-    );
 
     if (!validArgs) {
       throw new Error("INVALID ARGUMENTS");
     }
 
     return new Promise((resolve, reject) => {
-      args.push((err: Error | null, conn: any) => {
+      const callback = (err: Error | null, conn: any) => {
```
Comment (pontusvision) [inline src/drivermanager.ts]: remove callback?

File: src/drivermanager.ts [hunk 18]
```diff
@@ -1,93 +1,132 @@
...
     });
   },
 
   async setLoginTimeout(seconds: number): Promise<boolean> {
     return new Promise((resolve, reject) => {
-      java.callStaticMethod(DM, 'setLoginTimeout', seconds, (err: Error | null) => {
-        if (err) reject(err);
-        else resolve(true);
-      });
+      java.callStaticMethod(
```
Comment (pontusvision) [inline src/drivermanager.ts]: review if needed.

File: src/index.ts [hunk 19]
```diff
@@ -7,3 +7,85 @@ export { default as ResultSet } from './resultset';
 export { default as ResultSetMetaData } from './resultsetmetadata';
 export { default as SQLWarning } from './sqlwarning';
 export { default as Statement } from './statement';
+
+// export const classPath = process.env['CLASSPATH']?.split(',');
```
Comment (pontusvision) [inline src/index.ts]: remove; add in the readme / test cases

File: src/connection.ts [hunk 20]
```diff
@@ -1,111 +1,601 @@
...
+      resolve(this._conn?.closeSync())
+    });
   }
 
-  async clearWarnings(): Promise<void> {
-    await new Promise<void>((resolve, reject) => {
-      this._conn.clearWarnings((err: Error | null) => {
+  commit(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this._conn?.commit((err: Error | null) => {
```
Comment (pontusvision) [inline src/connection.ts]: use commitSync()

File: src/connection.ts [hunk 21]
```diff
@@ -1,111 +1,601 @@
...
+
+  createSQLXML(): Promise<void> {
+    return Promise.reject(new Error("NOT IMPLEMENTED"));
+  }
+
+  createStatement(arg1?: number, arg2?: number, arg3?: number): Promise<any> {
+    return new Promise((resolve, reject) => {
+      try {
+        const args = [arg1, arg2, arg3].filter((arg) => arg !== undefined); // Filter out undefined arguments
+        const statement = this._conn?.createStatement(...args); // Invoke without callback
```
Comment (pontusvision) [inline src/connection.ts]: Use Sync()?

File: src/connection.ts [hunk 22]
```diff
@@ -1,111 +1,606 @@
...
+const java = Jinst.getInstance();
+
+if (!java.isJvmCreated()) {
+  java.addOption("-Xrs");
 }
 
-export class Connection {
-  private _conn: any;
-  private _txniso: string[];
+type ConnectionType = {
```
Comment (pontusvision) [inline src/connection.ts:14]: Please remove this type

File: src/pool.ts [hunk 23]
```diff
@@ -1,180 +1,274 @@
...
-    keepalive: ka.enabled ? setInterval(() => keepalive(conn, ka.query), ka.interval) : false
+    keepalive: ka.enabled
+      ? setInterval(keepalive, ka.interval, conn, ka.query)
+      : false,
+      lastIdle: undefined
   };
 
   if (maxIdle) {
-    connObj.lastIdle = new Date().getTime();
+    connobj["lastIdle"] = new Date().getTime();
```
Comment (pontusvision) [inline src/pool.ts]: revert to old code

File: src/connection.ts [hunk 24]
```diff
@@ -1,111 +1,582 @@
...
+    return this.isReadOnlySync()
   }
 
-  async abort(executor: any): Promise<void> {
-    throw new Error("NOT IMPLEMENTED");
+  isValidSync(val: number):boolean {
+    return this.isValidSync(val)
+  }
+  abort(executor: any): Promise<void> {
+    return Promise.reject(new Error("NOT IMPLEMENTED"));
```
Comment (pontusvision) [inline src/connection.ts]: please implement.

File: src/pool.ts [hunk 25]
```diff
@@ -1,180 +1,279 @@
...
+  uuid: string; closed: boolean; readonly?: boolean; valid?: boolean
+}
 
-import _ from 'lodash';
-import { v4 as uuidv4 } from 'uuid';
-import jinst from "./jinst";
-import dm from './drivermanager';
-import Connection from './connection';
-import winston from 'winston';
+export interface ConnObj {
```
Comment (pontusvision) [inline src/pool.ts]: Please rename to PoolConnObj

File: src/pool.ts [hunk 26]
```diff
@@ -1,180 +1,279 @@
...
+    const status: PoolStatus = {};
+    status.available = this._pool.length;
+    status.reserved = this._reserved.length;
+    status.pool = this.connStatus([], this._pool);
+    status.rpool = this.connStatus([], this._reserved);
     return status;
   }
 
-  private async _addConnectionsOnInitialize(): Promise<void> {
+  private connStatus(acc: ConnStatus[], pool: PoolConnStatus[]): ConnStatus[] {
```
Comment (pontusvision) [inline src/pool.ts]: Please check this...  do we need await?

File: src/connection.ts (hunk 24 above)
Comment (pvpmartins) [reply src/connection.ts]: Parece que não há implementação para esse metodo em javascript:

```
Error aborting connection: Error: Error running instance method
java.sql.SQLFeatureNotSupportedException: Method not supported
```

Comment (pvpmartins) [reply src/connection.ts]: Quando tento remover o argumento que deveria ser de type Executor, obtenho:

Error executing query: Error: Could not find method "abort()" on class "class org.apache.hive.jdbc.HiveConnection". Possible matches:
  public void org.apache.hive.jdbc.HiveConnection.abort(java.util.concurrent.Executor) throws java.sql.SQLException
  
 Parece que não há implementação equivalente a este objeto em javascript. Isso que diz o ChatGPT:

"In TypeScript (or JavaScript), there's no direct equivalent to Java's Executor interface, as JavaScript has a single-threaded runtime and doesn’t use threads in the same way Java does."

**Review by pontusvision — APPROVED** [review]

**Pull request merged**